kubevirt provider e2e create cluster test

davidvossel opened the pull request:
This functional e2e test is very basic and is primarily meant as an entrypoint for us to begin fleshing out more advanced validation in the future. At the moment, we simply need a test case that we can wire up to Openshift CI to validate that a hosted cluster actually boots KubeVirt VMs successfully.

At the moment this test does the following.
- Posts a HostedCluster using the kubevirt provider
- Verifies the NodePool gets created as expected
- Verifies the KubeVirt Machines come online (meaning that the VMs boot, that's it)
- Verifies the KubeVirtCluster is ready

Unlike the AWS create cluster test, access to the guest cluster is not being exercised yet, and we aren't even verifying that the HostCluster reaches a completed state. This verification for the KubeVirt provider will come in time.

sjenning commented:
/test e2e-aws-all
to hit this test

davidvossel commented:
> /test e2e-aws-all
> to hit this test

@sjenning We actually need this test to be skipped on AWS. The issue here is that KubeVirt requires nested virtualization, and the non-metal AWS EC2 instances used by OpenShift CI do not support nested virtualization.

@nunnatsa is in the process of preparing such a lane on Azure which does support nested virtualization, but this will take time. 

Ideally, what I'd like to see is for this e2e test to get merged before a CI lane executes it as a way for @nunnatsa to have a simple e2e test to prove out the Azure lane works with kubevirt and nested virt. I'm sure we'll iterate on this test a few more times as that Azure lane is being shored up, but this at least gives us a foundation to work with.

Does that sound reasonable? These new test lanes are always kind of a pain due to the chicken and egg problem. We need a test exercise the new lane's abilities, but the test can't be proven until the lane exists. :/

sjenning commented:
So we need to exclude this test from e2e-aws-all then?
https://github.com/openshift/release/blob/master/ci-operator/config/openshift/hypershift/openshift-hypershift-main.yaml#L126

There will actually be no CI test that will execute this test this until the Azure work is done and... hypershift supports Azure which it currently does not?

alvaroaleman commented on test/e2e/create_cluster_test.go:
```diff
@@ -56,6 +56,47 @@ func TestAWSCreateCluster(t *testing.T) {
+//
+// This test is meant to provide a first, fast signal to detect regression; it
+// is recommended to use it as a PR blocker test.
+func TestKubeVirtCreateCluster(t *testing.T) {
```
Could we document what expecations for the management cluster we have so this works? I.E. how do you install kubevirt (is it installed by default in OCP?) and that it needs nodes that support KVM?

enxebre replied:
Agreed, We need a commit/separate PR to properly introduce kubevirt docs ([same for agent platform](https://github.com/openshift/hypershift/pull/808#issuecomment-1004216829))

davidvossel replied:
> Could we document what expecations for the management cluster we have so this works? I.E. how do you install kubevirt (is it installed by default in OCP?) and that it needs nodes that support KVM?

yep, @nirarg I believe this is something you're working on?

enxebre commented:
>There will actually be no CI test that will execute this test this until the Azure work is done and... hypershift supports Azure which it currently does not?

@sjenning Azure here is an arbitrary platform for the management cluster they choose to setup nested virt to accommodate Kubevirt. So then we can have a CI job which gives a management cluster with nested virt setup support (it'll happen to be in Azure but might be anywhere) so we can test hypershift platform=Kubevirt. This is not tied to hypershift supporting Azure. Make sense? cc @davidvossel 

>Does that sound reasonable? These new test lanes are always kind of a pain due to the chicken and egg problem. We need a test exercise the new lane's abilities, but the test can't be proven until the lane exists. :/

@davidvossel this sounds reasonable to me, if everyone else agrees I think we can merge after addressing comments and getting docs in place so we can enable @nunnatsa to follow up.

enxebre commented on test/e2e/e2e_test.go:
```diff
@@ -56,6 +56,8 @@ func TestMain(m *testing.M) {
 	flag.StringVar(&globalOpts.configurableClusterOptions.Region, "e2e.aws-region", "us-east-1", "AWS region for clusters")
 	flag.StringVar(&globalOpts.configurableClusterOptions.PullSecretFile, "e2e.pull-secret-file", "", "path to pull secret")
 	flag.StringVar(&globalOpts.configurableClusterOptions.AWSEndpointAccess, "e2e.aws-endpoint-access", "", "endpoint access profile for the cluster")
+	flag.StringVar(&globalOpts.configurableClusterOptions.KubeVirtContainerDiskImage, "e2e.kubevirt-container-disk-image", "", "container disk image to use for kubevirt nodes")
```
The existing flags here are environmental config.
Why are we making these flags? I see them as test case specific values, I'd rather keep them hardcoded until there's a reason to not.

enxebre commented on test/e2e/create_cluster_test.go:
```diff
@@ -56,6 +56,47 @@ func TestAWSCreateCluster(t *testing.T) {
+	e2eutil.WaitForHostedClusterAvailable(t, testContext, client, hostedCluster)
+
+	// Wait for kubevirt machines to come online
+	e2eutil.WaitForKubeVirtMachines(t, testContext, client, hostedCluster, *nodepool.Spec.NodeCount)
```
why is this not using `WaitForNReadyNodes`? if there's a reason please add a TODO

davidvossel replied:
Yep, i'll write a comment for why this is necessary.

The issue at the moment is that we don't have guest cluster access because there's no proper ingress. We're working on sorting out options for ingress at the moment. Right now the best we can do to get some preliminary test results of the KubeVirt platform is to introspect the HostedCluster's related CRs on the management cluster. It's not great, but gives us some confidence that at least the KubeVirt VMs launch and that the HostedCluster is marked as being available.

enxebre commented on test/e2e/util/util.go:
```diff
@@ -195,6 +201,100 @@ func WaitForGuestKubeConfig(t *testing.T, ctx context.Context, client crclient.C
 	return data, nil
 }
 
+func WaitForKubeVirtMachines(t *testing.T, ctx context.Context, client crclient.Client, hostedCluster *hyperv1.HostedCluster, count int32) {
```
can you please move this into a kubevirt file?

davidvossel replied:
done

enxebre commented on test/e2e/util/util.go:
```diff
@@ -195,6 +201,100 @@ func WaitForGuestKubeConfig(t *testing.T, ctx context.Context, client crclient.C
+	t.Logf("KubeVirtMachines are ready in %s", time.Since(start).Round(time.Second))
+}
+
+func WaitForKubeVirtCluster(t *testing.T, ctx context.Context, client crclient.Client, hostedCluster *hyperv1.HostedCluster) {
```
can you please move this into a kubevirt file?

davidvossel replied:
yep, no problem

enxebre commented:
Also you'll need to ensure this Test* only run against your management cluster see docs/content/contribute/run-tests.md

davidvossel commented:
I think I've addressed the feedback so far. Understand that this test is suboptimal. It's the highest bar I could come up with to validate the kubevirt platform given what exists right now. This will continue to improve in the future.

> Also you'll need to ensure these TestKubeVirt* run only against your management cluster see docs/content/contribute/run-tests.md

yep, we'll filter by a KubeVirt regex in our CI lane.

> So we need to exclude this test from e2e-aws-all then?

@sjenning I added usage of an environment variable `KUBEVIRT_PLATFORM_ENABLED=true` that indicates if the KubeVirt platform related tests can run. By default, unless that env var is set, the KubeVirt tests will be skipped. From what I can see, there's not a great way to tell go test to "run all tests except this subset" without using something like an env var. 

This preserves the usage of the ci lanes that run on AWS infra.

sjenning commented:
/retest

rmohr commented:
> So we need to exclude this test from e2e-aws-all then? https://github.com/openshift/release/blob/master/ci-operator/config/openshift/hypershift/openshift-hypershift-main.yaml#L126
> 
> There will actually be no CI test that will execute this test this until the Azure work is done and... hypershift supports Azure which it currently does not?

@sjenning that should be fine. We don't need hypershift to support azure. We just need "a" openshift cluster from azure (not hypershift) where we can deploy hypershift and kubevirt.

rmohr reviewed (commented):
lgtm from a KubeVirt perspective.

davidvossel commented on test/e2e/util/cluster/kubevirt/cluster.go:
```diff
@@ -0,0 +1,46 @@
+func (k *KubeVirt) CreateCluster(ctx context.Context, hc *hyperv1.HostedCluster) error {
+	k.opts.Namespace = hc.Namespace
+	k.opts.Name = hc.Name
+	//TODO add kubevirt options here.
```
removed... Originally i thought that's where i needed to fill in the KubeVirt platform related options, but instead i needed to do it in the global default e2e config in another file.

nunnatsa commented on test/e2e/util/kubevirt.go:
```diff
@@ -0,0 +1,94 @@
+
+// KubeVirtEnabled indicates if the test environment is
+// setup to allow usage of the KubeVirt platform.
+func KubeVirtEnabled() bool {
```
What will set this env var? The test?

Maybe we can read the KV CRDs instead?

davidvossel replied:
We'll set it in the openshift ci config for the kubevirt test lane.

I want the test to be explicitly enabled or disabled via an env var in order to avoid false positive test results that simply silently skipped the kubevirt tests. This is based on experience. I've seen where we've attempted to create auto detection logic to determine when a test should run or not cause issues (unexpectedly skipped) which allowed regressions

enxebre replied:
I don't think we shouldn't introduce an arbitrary pattern that rely on each single test checking an env variable.
Atm these are developed as regular go tests, the built in way to run specific tests is via go test -run "prefix" which should be ensure by the caller, e.g ci bin/test-e2e -test.v -test.timeout 0 -test.run='^TestAgent.*'.

davidvossel replied:
what you suggest breaks the aws run-all lane. We'd have to modify that test lane to opt in to specific tests, or we can go with the env var solution I'm proposing here which skips tests which require different environmental setup.

I'm fine with either direction. Which one makes the most sense for you all?

ironcladlou replied:
I don't see an issue with dividing our test configs in CI right now and relying on `-test.run` to whitelist the suite for the given CI test definition. It's not clear we want all these tests in the blocking presubmits, for example. Maybe I'm misunderstanding the problem still.

enxebre replied:
Another thinking I'm having now while looking at our code is all platforms should aim to pass this tests. We could keep tests names platform agnostic TestUpgradeControlPlane, TestAutoRepair...
Do the platform discrimination within a single func e.g hostedCluster := e2eutil.Create(Platform)Cluster based on whatever platorm the test is running on which can be passed through t parameter. Your e2e job will fail until is able to not, which makes sense. Thoughts?

ironcladlou replied:
I'm with you, I don't think we can address it in the scope of this PR though. I started taking some steps in the general direction with https://github.com/openshift/hypershift/pull/884 (which unwinds most of the abstractions). Let's take the broader internal arch discussion elsewhere?

enxebre replied:
ok @ironcladlou I see that's pretty much what you did here https://github.com/openshift/hypershift/pull/884/files

ironcladlou replied:
Our periodics currently run all the tests in the e2e package and don't use `-test.run` to filter. So either that has to change (and the test names need updated so we can whitelist them), or we need another strategy to prevent this test from running there (which assumes AWS).

ironcladlou replied:
My suggestion for now is to just move the env var test directly into TestKubeVirtCreateCluster. There's only one WIP test now, so I don't see the need to extract a whole utility file and codify it at this point.

davidvossel replied:
i moved the env var check to the kubevirt test and left a TODO note about the long term direction we're shooting for here.

enxebre commented on test/e2e/util/cluster/kubevirt/cluster.go:
```diff
@@ -0,0 +1,45 @@
+}
+
+func (k *KubeVirt) DumpCluster(ctx context.Context, hc *hyperv1.HostedCluster, artifactsDir string) {
+	aws.DumpHostedCluster(k.t, ctx, hc, artifactsDir)
```
we might want to make this generic https://github.com/openshift/hypershift/pull/877.

davidvossel replied:
this is shared with the existing  `none` cluster logic as well

ironcladlou commented:
Hi folks, I want to help get this through, but the current factoring of the e2e code with regards to platform specific handling is not satisfactory to @alvaroaleman or myself. We have discussed how we'd rather handle it, and need to do some refactoring in here. The way this PR is currently factored seems to further expand and entrench a pattern I'm not sure we want. Can we please have a discussion about this before moving forward? I am working on a WIP to bring the code in line with what @alvaroaleman had briefly discussed.

I want to avoid wasting time working at cross purposes.

ironcladlou commented:
https://github.com/openshift/hypershift/pull/884 is where I'm going, but I don't want to make you rebase. I'm fine to let this through and then I can rebase.

ironcladlou commented on test/e2e/create_cluster_test.go:
```diff
@@ -7,9 +7,64 @@ import (
+	t.Logf("Created nodepool. Namespace: %s, name: %s", nodepool.Namespace, nodepool.Name)
+
+	// Wait for hosted cluster to become ready
+	e2eutil.WaitForHostedClusterAvailable(t, testContext, client, hostedCluster)
```
i.e. if this is just temporary because the test coverage is incomplete, please add a comment/TODO

davidvossel replied:
yes, this is a work around. I've refactored it as such.

I moved the until function into this test so it won't be seen as a generic function other tests should adopt. I left a TODO note in the test to transition to the WaitForImageRollout once the full image roll out works consistently for us.

At the moment, the bar for our e2e verification is basically just that a hosted cluster with KubeVirt platform is created, and KubeVirt VMs actually launch successfully hitting a `ready` state. We've got a lot of work to do to mature the kubevirt platform :)

davidvossel commented:
/hold

I want to test my changes again here locally before this merges. Feel free to lgtm/approve in the mean time if someone feels comfortable and I'll remove the hold once I verify it's all still in order from my end.

> I'm fine to let this through and then I can rebase.

Thanks, I followed similar patterns to aws and none platforms, so hopefully it's not much of an inconvenience.

ironcladlou commented:
/approve
/lgtm

No further feedback on my end for now, happy to get the skeleton in place for continued iteration

davidvossel commented:
/hold cancel

Passes locally now. I had to make one minor change.

@ironcladlou For the test to pass, I had to make a change and add the capk scheme to the test client because i'm doing some introspection of the KubeVirtMachine and KubeVirtCluster objects.

ironcladlou commented:
> /hold cancel
> 
> Passes locally now. I had to make one minor change so this will need another `lgtm`
> 
> @ironcladlou For the test to pass, I had to make a change and add the capk scheme to the test client because i'm doing some introspection of the KubeVirtMachine and KubeVirtCluster objects.

Awesome, thanks!

/lgtm

davidvossel commented:
> @davidvossel sorry I didn't mention this before, but can you please squash the commits?

yep, squashed

ironcladlou commented:
/lgtm

davidvossel commented:
@nunnatsa  fyi, here's an example of how I ran this test locally from the source tree. 

```
export KUBEVIRT_PLATFORM_ENABLED=true
make e2e
bin/test-e2e \
          -test.v \
          -test.timeout=0 \
          -test.run=TestKubeVirtCreateCluster \
          -e2e.node-pool-replicas=1 \
          -e2e.base-domain="fake" \
          -e2e.kubevirt-container-disk-image="$RHCOS_CD_IMAGE" \
          --e2e.pull-secret-file="$PULL_SECRET_PATH"
```
